### ISSUE TITLE
Ingestor is no more Dataset ingestor

### ISSUE THREAD

**Issue by jkotan** [SciCatProject/backend#712]
I would assume the main metadata class in SciCat is Dataset i.e. it describes the most inportant details of scans or runs. Therefore `Ingestor` should have possiblity to ingest Datasets. Unfortunately, after
  https://github.com/SciCatProject/scicat-backend-next/pull/415/ where the to following lines where commented
```patch
--- a/src/casl/casl-ability.factory.ts
+++ b/src/casl/casl-ability.factory.ts
@@ -215,8 +215,8 @@ export class CaslAbilityFactory {
       can(Action.Create, Attachment);
 
       //cannot(Action.Delete, DatasetClass);
-      can(Action.Create, DatasetClass);
-      can(Action.Update, DatasetClass);
+     //can(Action.Create, DatasetClass);
+     //can(Action.Update, DatasetClass);
 
       can(Action.Create, Instrument);
       can(Action.Update, Instrument);

```
the ingestor is allowed to ingest only Instrument metadata unless it ingestor = ownerGroup or ingestor is in ADMIN_GROUPS.

At DESY in the old backend version we use ingestor account to ingest Dataset metadata where ownerGroup is our beamtime user group (different from ingestor). Also we don't want to set admin privileges to ingestor.

Therefore I would like to ask what was the reason to comment those two lines out and if it possible to revert the changes, i.e.
make `ingestor` back `the ingestor`. 

If uncommenting the above lines contradicts other use-cases I would propose to define a new role e.g. datasetIngestor 
which has permissions to create/update Datasets, i.e. similar to proposalIngestor for Proposals (This solution was originally proposed by Linus).
Other option would be to create a new env variable, e.g. CREATE_DATASET_PRIVILEGED_GROUP to switch the privileges (proposed by Max) . However, the latter solution look to me less readable.

**Comment by nitrosx** [comment]
@jkotan thank you for opening the issue.
In the new backend, we are moving away (slowly) from roles and permissions are group based.
You can change the casl-ability-factory back, but that means that only users belonging to the role/group _ingestor_ will be able to create datasets with no limitations.
My proposed solution allows institutes to easily configure which group is allowed to performed such action, and the group can be different from _ingestor_ . Think about when an institution has to use a predefined LDAP user for ingestion. Also you could configure multiple groups with ingestion permissions.
My intention was to make it easier for non-expert users to configure permissions and provide something like a framework for the permissions.

That said, potentially you could rewrite completely the casl-ability file to match exactly your needs. The only draw back is that you have to rebuilt the new docker image on your own.

Please let me know your thoughts

**Comment by jkotan** [comment]
Hello @nitrosx, thank you for your explanations. 

So it looks like the old mechanizm with the `ingestor` role  for ingesting dataset was commented out and the new mechanizm  with  CREATE_DATASET_PRIVILEGED_GROUP has been not implemented yet.

If I understand you correctly you would like to still use roles but in a hidden way i.e. not naming them roles but defines  them via ENV variables which become less clear to me who is whom and what they can do.

A  possibility of define who is the ingestor from LDAP users which is quite nice idea. 
However, currently for DESY  it is enough we have one functional account to ingest datasets which is not ownerGroup i.e. we encode beamtime/beamline in the dataset pid so for us it is good the ingestion processes it automatize so we can set correct dataset pid to exclude possibility one beamline user interferes with other beamline datasets. For us It is sufficient the owerGroup has only possibility to update metadata of their existing datasets.

This ingestor account could be also defined from LDAP users but using keycloak authentication it is hard to get token from a level of python code. Probably we should switch to use tokens directly. On the other side you need to change token much more frequently than passwords and it could be annoying on more distributed systems (but do-able).

Also your words "we are moving away (slowly) from roles and permissions are group based" scares me a little bit because sounds  like a continuous process of braking backward compatibility, reducing permissions more and more from one release to another (but I hope you don't mean that :) )

**Comment by nitrosx** [comment]
@jkotan 

> So it looks like the old mechanizm with the ingestor role for ingesting dataset was commented out and the new mechanizm with CREATE_DATASET_PRIVILEGED_GROUP has been not implemented yet.   

You are correct. I'm planning to work on it next week. IT should not be a huge effort

> If I understand you correctly you would like to still use roles but in a hidden way i.e. not naming them roles but defines them via ENV variables which become less clear to me who is whom and what they can do.    

I'm getting rid of roles entirely. All access is controlled by groups. Any role left over is due to lack of time during the migration.
Functional accounts that still have roles defined, are converted to groups.
For example, the ingestor account which previously had role ingestor, it will now belong to group ingestor. The conversion is done automatically. I have decided to implement the translation this way in order to be backward compatibile

> Also your words "we are moving away (slowly) from roles and permissions are group based" scares me a little bit because sounds like a continuous process of braking backward compatibility, reducing permissions more and more from one release to another (but I hope you don't mean that :) )   

I agree with you!!! I used the word _slowly_ ironically to express my utter frustration on how slow I have been able to transition away from roles.

**Comment by bpedersen2** [comment]
I think this is long fixed? @nitrosx

**Comment by jkotan** [comment]
It is not fixed yet. ingestor without ADMIN permissions cannot check if a dataset exists or read a dataset to patch it

**Comment by jkotan** [comment]
Hello @nitrosx  and @bpedersen2, is there any new way to add read dataset permissions for ingestor?

**Comment by nitrosx** [comment]
@jkotan I need to review the code, before I can give you a definitive answer.
Just to be sure that I understand: you would like to have an account named "ingestor" that can create and also update datasets that belongs to a group which he/she is not part of it? correct?
Thanks

**Comment by jkotan** [comment]
Hi @nitrosx , the ingestor is our functional account to ingest datasets. At DESY in our workflow 
We do not allow our users to create datasets.
1) the ingestor checks if dataset with given pid exists
2) if not exists create the new datasets
3) if exists  the ingestor reads and patches the existing dataset
Currently  1)  and 3) we do by putting  ingestor to `ADMIN_GROUPS`. The question is if it is possible to do it without it?

**Comment by nitrosx** [comment]
Thanks for the explaining.
I need to refresh my memory, than I will provide the answer.

**Comment by nitrosx** [comment]
@jkotan after review the code, I can confirm that at the moment it is not possible to assign elevated privileges to a group of users to just update existing datasets, unless you assign them to the ADMIN_GROUPS.
I propose to create UPDATE_DATASET_PRIVILEGED_GROUPS, which will allow the users of the groups listed to have elevated update privileges.
With this solution, you would need to place your ingestor group in the CREATE_DATASET_PRIVILEGED_GROUPS and in UPDATE_DATASET_PRIVILEGED_GROUPS to achieve your needs.

**Comment by jkotan** [comment]
Hi @nitrosx . Thanks. Is it possible to add also  READ_DATASET_PRIVILEGED_GROUPS ? I don't like idea to put ingestor function user to each access group.

**Comment by nitrosx** [comment]
Good call!!!
